### PR TITLE
PFM-ISSUE-6660 - Handle LESS escaping

### PR DESCRIPTION
**Pull request by slaven3kopic**
resolves [PFM-ISSUE-6660](https://base.cplace.io/pages/g4eu5s15zme2c5p54jb3rmdat/PFM-ISSUE-6660-Handle-LESS-escaping#/__8jj9zuc561yf9hivrckouo7nl/fe2.0/cf-cplace-platform/0.1.0/)

### NOTE
- to avoid cplace to stop working, only a warning will be displayed. Error will not be thrown.
- warnings in the main repository are fixed with this [PR](https://github.com/collaborationFactory/cplace/pull/4830)

### How to test
- run `npm link` in the cplace-asc repository
- run `cplace-asc` in the cplace repository

#### Expected result

Screen Shot 2021-05-20 at 15 19 54

### CODE REVIEW

File: src/compiler/LessCompiler.ts [hunk 1]
```diff
@@ -113,6 +119,25 @@ export class LessCompiler implements ICompiler {
...
     }
 
+    /**
+     * Checks improper LESS escaping
+     * @param cssOutput CSS string output
+     * @private
+     */
+    private checkImproperLessEscaping(cssOutput: string): string | undefined {
+        const cssOutputArray = cssOutput.split('\n');
+        const foundImproperLessEscaping = cssOutputArray.reduce((acc: string[], val: string) => {
```
Comment (Seledrex) [inline src/compiler/LessCompiler.ts:129]: This is making warnings for all `calc` expressions, whether they are escaped or not. The css output will not have the escaping included I believe.

Comment (slaven3kopic) [reply src/compiler/LessCompiler.ts:129]: True. I am on it. It was my bad assumption. Thanks for the hint 👍

**Comment by rishoocplace** [comment]
Tested on windows, works good.
LGTM 👍 
Already discussed this [observation ](https://base.cplace.io/pages/ibhn67vw1tmz3zqxox91iqior/The-hint-message-can-be-displayed-once-for-each-plugin?highlightedEntityUid=page%2Fibhn67vw1tmz3zqxox91iqior) with @slaven3kopic and this is not needed.

File: src/model/LessEscapePlugin.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ function checkImproperLessEscaping(cssOutput: string): string | undefined {
     }
 }
 
-let processedFiles: string[] = [];
+let processedFiles: Set<string> = new Set<string>();
```
Comment (Seledrex) [inline src/model/LessEscapePlugin.ts:22]:
```suggestion
let improperLessEscapedFiles: Set<string> = new Set<string>();
```

File: src/model/LessEscapePlugin.ts [hunk 3]
```diff
@@ -29,14 +29,18 @@ export function lessEscapePlugin(pluginName: string) {
...
+                    cwarn`\n${improperLessEscaping}\n\n`,
+                    cwarn`${spacer}Please escape LESS the following way:\n`,
+                    cwarn`${spacer}width: ~"calc(100% - 200px)"; or width: ~"calc(100vw - " @yourVariable ~")";\n`
+                );
+                processedFiles.add(extra.fileInfo.filename);
             } else {
-                const foundFile = processedFiles.indexOf(extra.fileInfo.filename);
-                if (foundFile !== -1) {
-                    processedFiles.splice(foundFile, 1);
+                if (processedFiles.has(extra.fileInfo.filename)) {
```
Comment (Seledrex) [inline src/model/LessEscapePlugin.ts:42]: Would be nice to add comment about watch mode here

File: src/compiler/LessCompiler.ts [hunk 4]
```diff
@@ -47,7 +48,8 @@ export class LessCompiler implements ICompiler {
         console.log(`⟲ [${this.pluginName}] starting LESS compilation...`);
         return new Promise<CompilationResult>((resolve, reject) => {
             const lesscOptions: Options = {
-                filename: path.resolve(entryFile)
+                filename: path.resolve(entryFile),
+                plugins: lessPlugins(this.pluginName)
```
Comment (smasala) [inline src/compiler/LessCompiler.ts:52]: 👍

**Review by smasala — APPROVED** [review]

**Review by Seledrex — APPROVED** [review]
Awesome! Thanks for the great work!

**Pull request merged**